# Herramientas GNU/Linux

Este documento muestra el uso de los comandos **GREP, SORT, CUT, TR, UNIQ, JOINT, PASTE, TAIL, HEAD, ECHO** y el **PIPE** "__|__".

## -GREP-

Se emplea para la búsqueda de texto/patron en un archivo.

Sintáxis básica: **grep** *cadena_texto   archivo*

### Ejemplo 

Para extraer de un archivo los lineas con el patrón de estación "BJU" se puede emplear el comando siguiente:

```
grep BJU lista.dat
```

Para extraer ahora la vuariable PMCO de ese mismo archivo se puede emplear:


In [5]:
%%bash -l

grep PMCO lista.dat

01/01/2021 01:00,PMCO,MGH
01/01/2021 01:00,PMCO,AJM
01/01/2021 01:00,PMCO,MPA
01/01/2021 01:00,PMCO,BJU
01/01/2021 01:00,PMCO,INN
01/01/2021 01:00,PMCO,GAM
01/01/2021 02:00,PMCO,CAM
01/01/2021 02:00,PMCO,BJU
01/01/2021 02:00,PMCO,MGH
01/01/2021 03:00,PMCO,BJU
01/01/2021 04:00,PMCO,BJU


### Ejemplo

Cuando se requieran extraer las líneas que contengan, o el patrón de estación __CUA__ o el patron con la variable __PMCO__, se necesita definir cada patrón mediante el patrón (-e). Además, si se requiere que se ignore si son mayúsculas o minúsculas se emplea el parámetro (-i).

Definición:

-e patrón

-i ignorar mayúsculas y minúsculas

```
grep  -i -e cua -e pmco lista.dat
```

In [3]:
%%bash -l

grep -i -e cua -e pmco lista.dat

01/01/2021 01:00,PMCO,MGH
01/01/2021 01:00,PMCO,AJM
01/01/2021 01:00,PMCO,MPA
01/01/2021 01:00,PMCO,BJU
01/01/2021 01:00,PMCO,INN
01/01/2021 01:00,PMCO,GAM
01/01/2021 02:00,PMCO,CAM
01/01/2021 02:00,CO,CUA
01/01/2021 02:00,NO,CUA
01/01/2021 02:00,NO2,CUA
01/01/2021 02:00,PMCO,BJU
01/01/2021 02:00,PMCO,MGH
01/01/2021 03:00,PMCO,BJU
01/01/2021 04:00,PMCO,BJU


### Ejemplo

Si se necesita que sean exclusivas, es decir que las líneas extraidas tengan ambos patrones se puede emplear el comando pipe "__|__" para realizar la segunda búsqueda (o filtrado).

```
grep -i -e BJU lista.dat | grep PMCO
```

o bien gregando -i en la segunda búsqueda si escribiste pmco en minúsculas. Ambas salidas son equivalentes.

In [6]:
%%bash -l
grep -i -e bju lista.dat | grep -i pmco

01/01/2021 01:00,PMCO,BJU
01/01/2021 02:00,PMCO,BJU
01/01/2021 03:00,PMCO,BJU
01/01/2021 04:00,PMCO,BJU


***
## -SORT-

Ordena las líneas de un archivo de texto.

Sintáxis básica: **sort** *nombre_archivo*

### Ejemplo
```
sort lista.dat
```

### Ejemplo

Si se deseara ordenar el archivo *lista.dat* mediante la columna 3 sabiendo que el separador de cada columna son las comas se puede emplear lo siguiente:

```
sort -k3 -t"," lista.dat > lista_Est.txt
```

Definición:

-k especifica la columna para realizar el ordenamiento.

-t especifica el separador para identificar las columnas.

Además, en esta instrucción la salidaa se direccionana a un archivo con nombre *lista_Est.txt* mediante el redireccionamiento con "__>__". Si esta parte se omitiera la salida solo se presentaría en la pantalla.

```
sort -k3 -t"," lista.dat
```

In [32]:
%%bash -l

sort -k3 -t"," lista.dat > lista_Est.txt

El archivo quedará ordenado y puedes visualizarlo abriendo una **Terminal** desde el *Launcher* mediante los comandos **cat** o **more** (vistos en la primera clase: https://tlaloc.atmosfera.unam.mx/cca-ucar/curso_unix/intro/ ).

O bien puedes verificar el resultado viendo el principio y la parte final del archivo mediante **head** y **tail**.

***
## -Head-

Imprime las primeras n líneas superiores de un archivo (por defecto n=10). Si se quiere manipular el número de líneas a imprimir se usa con la opción -n *numero_lineas*.

Sintáxis básica: **head** nombre_archivo

### Ejemplos
```
head lista.dat
```

```
head -n 7 lista.dat
```
***
## -Tail-

Sintáxis básica: **tail** nombre_archivo

Imprime las últimas n líneas de un archivo (por defecto n=10). Si se quiere manipular el número de líneas a imprimir se usa con la opción -n *numero_lineas*.

### Ejemplos
```
tail lista.dat
```

```
tail -n 14 lista.dat
```

Ahora verifiquemos el inicio y el fin de ambos archivos: lista.dat (archivo origianl sin ordenar) y lista_Est.txt (archivo ordenado).

In [27]:
%%bash -l
# Archivo original
head -n 3 lista.dat 
echo "..."
tail -n 3 lista.dat
echo "Líneas en el archivo"
wc -l lista.dat
echo "          "

# Archivo ordenado
head -n 3 lista_Est.txt
echo "..."
tail -n 3 lista_Est.txt
echo "Líneas en el archivo"
wc -l lista_Est.txt

01/01/2021 01:00,NOX,ccA
01/01/2021 01:00,O3,CCA
01/01/2021 01:00,PM2.5,CCA
...
01/01/2021 04:00,PM2.5,BJU
01/01/2021 04:00,PMCO,BJU
01/01/2021 04:00,SO2,BJU
Líneas en el archivo
152 lista.dat
          
01/01/2021 02:00,CO,ACO
01/01/2021 02:00,CO,ACO
01/01/2021 02:00,NO2,ACO
...
01/01/2021 01:00,O3,SAC
01/01/2021 01:00,PM2.5,SAC
01/01/2021 01:00,SO2,SAC
Líneas en el archivo
152 lista_Est.txt


***
### Algunas precisiones antes de continuar

## -Echo-

Para mostrar líneas de texto/cadenas. Beben de estar entre comillas.

Sintáxis básica: **echo** *"lineas_texto"*

### Ejemplos
```
echo "Líneas en el archivo"
```

```
echo "..."
```
***
## -WC (word count)-

Empleado para contar líneas, palabras, caratéres y bytes en los archivos especificados.

Sintáxis básica: **wc** *nombre_archivo*

### Ejemplo
```
wc archivo.dat
```

Para este caso el resultado será: 

152  304 3807 lista.dat

Lo cual son 4 columnas, siendo la primer columna el número de líneas en el archivo, la segunda el número de palabras, la tercera el número de caracteres y la cuarta el nombre del archivo que se dió como argumento.

Para sólo imprimir el número de líneas que tiene el archivo se emplea junto con la opción -l

### Ejemplo
```
wc -l lista_Est.txt
```

**Ejercicio:** Ejecuta la instrucción para ordenar el archivo **lista.dat** mediante la columna 2 y guarda el resultado en el archivo **lista_comp.txt**






***
## -TR-

Se emplea para cambiar o remover caracteres en en un archivo.

### Ejemplo

Para substituir en un archivo todas las aparciones del caracter "," por tabuladores "\t" se puede realizar:
```
cat lista_Est.txt | tr "," "\t"  > lista_Est_tab.txt
```

o bien puede escribirse de la siguiente manera:
```
tr "," "\t" < lista_Est.txt > lista_Est_tab.txt
```

In [33]:
%%bash -l

tr "," "\t" <lista_Est.txt> lista_Est_tab1.txt

# para verificar el resultado
head -n 4 lista_Est_tab1.txt

01/01/2021 02:00	CO	ACO
01/01/2021 02:00	CO	ACO
01/01/2021 02:00	NO2	ACO
01/01/2021 02:00	NO2	ACO


para el caso en que se quiera substituir por espacios se puede emplear:
```
tr "," " " <lista_comp.txt> lista_comp_esp.txt
```
Nota que estamos trabajando sobre el archivo *lista_comp.txt* que se quedó como ejercicio anteriormente.

In [34]:
%%bash -l
tr "," " " <lista_comp.txt> lista_comp_esp.txt

# para revisar el cambio de coma a espacio
head -n 4 lista_comp_esp.txt

01/01/2021 02:00 CO ACO
01/01/2021 02:00 CO ACO
01/01/2021 01:00 CO AJM
01/01/2021 02:00 CO ATI


***
## -CUT-

Elimina fragmentos en cada línea de un archivo.

Sintáxis simple: ***cut** *opciones nombre_archivo*

### Ejemplo

Seleccionar la columna 2 del archivo _lista_comp.txt_ que se encuentra separado por comas:

```
cut -d "," -f 2 lista_comp.txt > lista_comp1.txt 

```
la opción -d especifique el delimitador que será considerado en lugar del tabulador "\t" que se encuentra como predeterminado. Mientras que la opción -f especifica el campo/columna a usar.


In [37]:
%%bash -l 
cut -d "," -f 2 lista_comp.txt > lista_comp1.txt

head -n 3 lista_comp1.txt
echo ...
tail -n 3 lista_comp1.txt
wc -l lista_comp1.txt

CO
CO
CO
...
SO2
SO2
SO2
152 lista_comp1.txt


In [39]:
%%bash -l 
# para especificar más de un campo
cut -d "," -f 2,3 lista_comp.txt > lista_comp2.txt

head -n 3 lista_comp2.txt
echo ...
tail -n 3 lista_comp2.txt
wc -l lista_comp2.txt

CO,ACO
CO,ACO
CO,AJM
...
SO2,MON
SO2,MPA
SO2,SAC
152 lista_comp2.txt


***
## -UNIQ-

Muestra u omite líneas duplicadas en un archivo.

### Ejemplo

Muestra los valores únicos en el archivo lista_comp1.txt 
```
uniq lista_comp1.txt
```
**NOTA**: Recuerda que este archivo sólo tiene una columna.

In [41]:
%%bash -l

uniq lista_comp1.txt > compuestos.txt

head compuestos.txt
# Cuenta el número de renglones
wc -l compuestos.txt

CO
NO2
NO
NOX
O3
PM10
PM2.5
PMCO
SO2
9 compuestos.txt


### Ejemplo

Para extraer cuántas estaciones hay (columna 3) del archivo original *lista.dat* se puede realizar lo siguiente:

```
cut -d "," -f 3 lista.dat |sort |uniq > estaciones_unicas.txt

```

In [43]:
%%bash -l

cut -d "," -f 3 lista.dat |sort |uniq > estaciones_unicas.txt

# Para comprobar:
head -n 3 estaciones_unicas.txt 
echo ...
tail -n 3 estaciones_unicas.txt
echo "numero de líneas en el archivo"
wc -l  estaciones_unicas.txt

ACO
AJM
AJU
...
MON
MPA
SAC
numero de líneas en el archivo
17 estaciones_unicas.txt


***
# JOIN

Une las líneas de dos archivos en función de un campo (columna) común presente en ambos archivos.

Sintáxis simple: **join** *opciones nombre_archivo1 nombre_archivo2*

### Ejemplo

Supongamos que se requiere agregar al archivo de las estaciones (**lista_Et.txt**) una descripción de cada estación (**des_estaciones.txt**). Sabiendo que tienen un campo (o columna) en común se hará la unión a través de este:
```
lista_Est.txt
01/01/2021 02:00,CO,ACO
01/01/2021 02:00,CO,ACO
01/01/2021 02:00,NO2,ACO

des_estaciones.dat
ACO,Acolman,19.635501,-98.912003
AJM,Ajusco Medio,19.272161,-99.207744
AJU,Ajusco,19.154286,-99.162611
ARA,Aragón,19.470218,-99.074549
ATI,Atizapan,19.576963,-99.254133
```
 
```
join -t "," -13 -21 lista_Est.txt des_estaciones.dat > list_Est_lo_lat.txt
```
donde la opción **-13** indica archivo 1 campo 3 y la **-21** indica archivo 2 campo 1. 

In [45]:
%%bash -l

join -t "," -13 -21 lista_Est.txt des_estaciones.dat > list_Est_lon_lat.txt

# Para revisar la lista
head -3 list_Est_lon_lat.txt
echo ...
tail -3 list_Est_lon_lat.txt

ACO,01/01/2021 02:00,CO,Acolman,19.635501,-98.912003
ACO,01/01/2021 02:00,CO,Acolman,19.635501,-98.912003
ACO,01/01/2021 02:00,NO2,Acolman,19.635501,-98.912003
...
SAC,01/01/2021 01:00,O3,Santiago Acahualtepec,19.34561,-99.009381
SAC,01/01/2021 01:00,PM2.5,Santiago Acahualtepec,19.34561,-99.009381
SAC,01/01/2021 01:00,SO2,Santiago Acahualtepec,19.34561,-99.009381


***
# PASTE

Une archivos horizontalmente generando líneas que consisten en líneas de cada arhivo de entrada y son separadas por tabuladores.

Sintáxis simple: **paste** *opciones nombre_archivo1 nombre_archivo 2*

### Ejemplo

Supongamos que se tiene el archivo de identificadores __id_est.txt__ y se desea acoplar con el archivo de localización __loc.txt__

```
id_est.txt     loc.txt
 
DIC             19.298819 -99.185774
EDL             19.313357 -99.310635
INN             19.291968 -99.38052
LPR             19.534727 -99.11772
MON             19.460415 -98.902853
```
para combinarlos o pegarlos se puede emplear:

```
paste id_est.txt loc.txt > combinado.txt
```


In [47]:
%%bash -l
# Combina con PASTE
paste id_est.txt loc.txt > combinado.txt

# Verifica
head combinado.txt

DIC	19.298819	-99.185774
EDL	19.313357	-99.310635
INN	19.291968	-99.38052
LPR	19.534727	-99.11772
MON	19.460415	-98.902853


***
## -GREP en archivos-

Es posible realizar una búsqueda de un patrón/texto en varios archivos.

Ahora con los archivos creados a partir de los ejemplos busquemos la palabra *ACO* en todos los archivos que inician con list. 

In [49]:
%%bash -l
# Busqueda de patrones en varios archivos

grep -l ACO list*

lista_comp2.txt
lista_comp_esp.txt
lista_comp.txt
lista.dat
lista_Est_tab1.txt
lista_Est_tab.txt
lista_Est.txt
list_Est_lon_lat.txt


***
Finalmente, si deseas borrar los archivos creados puedes emplear el comando **rm**